Skip Mode

wxtim opened the pull request:
Companion of https://github.com/cylc/cylc-flow/pull/6039

Closes #783 
Closes https://github.com/cylc/cylc-doc/issues/784

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.

oliver-sanders commented on src/user-guide/running-workflows/simulation-modes.rst:
```diff
@@ -3,8 +3,8 @@
 
-The run modes described in :ref:`task-run-modes` can
-be applied to the entire workflow.
+The "live", "simulation" and "dummy" modes described in
+:ref:`task-run-modes` can be applied to the entire workflow.
```
```suggestion
The "live" and "simulation" modes described in
:ref:`task-run-modes` can be applied to the entire workflow.
```

oliver-sanders commented on src/user-guide/task-implementation/run-modes.rst:
```diff
@@ -5,15 +5,24 @@ Task Run Modes
 
 .. versionadded:: 8.4.0
 
+<<<<<<< HEAD
```
:(

wxtim replied:
Rah!

oliver-sanders commented:
Please could you update the advice for family-family triggering to make use of skip mode.

https://cylc.github.io/cylc-doc/stable/html/workflow-design-guide/efficiency.html#family-to-family-triggering

Worth skimming the docs to see if we document any other dummy task use cases anywhere.

MetRonnie commented on src/user-guide/task-implementation/run-modes.rst:
```diff
@@ -0,0 +1,207 @@
+
+.. _task-run-modes.dummy:
+
+Dummy Mode
```
I don't think dummy and simulation should have been moved to a "task run mode" page?

wxtim replied:
Arguable, because they are still task modes, but I take your point, and will move them back.

wxtim commented:
> Please could you update the advice for family-family triggering to make use of skip mode.
> 
> https://cylc.github.io/cylc-doc/stable/html/workflow-design-guide/efficiency.html#family-to-family-triggering
> 
> Worth skimming the docs to see if we document any other dummy task use cases anywhere.

Good call: I've realized responding to Ronnie's comments that we've somewhat overloaded the term dummy to mean
* A task formally following the dummy task submission pathway
* An informal dummy mode task (i.e. an implicit task) or a task with `script = true`.
* A task which is a "dummy" with respect to it's graphing function. It can't be a true dummy task, but it could be a dummy like live task as described above, or a task carrying our some real work.

I've tried to remove either of the second two use cases for the word dummy.

I also found two references to "dummy-local" mode which hasn't been a thing for at least 3 years (I deleted it).

hjoliver commented:
> Good call: I've realized responding to Ronnie's comments that we've somewhat overloaded the term dummy to mean

Repeating my comment from the cylc-flow PR:

> IMO opinion a "dummy task" is just a task that has no purpose beyond its role in the graph structure, i.e. it does not run a job that does anything useful at all. So dummy mode and simulation mode (and now skip mode) both (all) run dummy tasks.
>
> However, it is potentially a bit confusing given the distinct mode names.
> 
> Now that we have (presumably) made simulation mode a lot more robust and useful ... can we consider ditching dummy mode? (we did discuss this somewhere once...). I'll put this question on the meeting agenda...
> 
> Either way, we should get the terminology straight and document it.

hjoliver commented on src/user-guide/task-implementation/run-modes.rst:
```diff
@@ -0,0 +1,211 @@
+   when developing a workflow, the scheduler simulates
+   running a task without creating a "real" job.
+#. :ref:`task-run-modes.dummy`: Intended to be used
+   when developing a workflow, the scheduler replaces
+   the workflow scripts with dummy scripts and returns
+   to the default ("live") submission pathway.
```
The submission "pathway" thing again - I think that kinda refers to internal implementation and might confuse users. Could be much simpler, e.g.:
```suggestion
   when developing a workflow. Instead of the real configured task scripts,
   the scheduler submits local jobs with empty (do nothing) "script" items.
```

hjoliver commented on src/user-guide/task-implementation/run-modes.rst:
```diff
@@ -0,0 +1,211 @@
+   the workflow scripts with dummy scripts and returns
+   to the default ("live") submission pathway.
+#. :ref:`task-run-modes.skip`: Intended to allow more control of the workflow,
+   the scheduler marks task outputs as completed.
```
```suggestion
   the scheduler marks task outputs as completed without submitting a job.
```

hjoliver commented on src/user-guide/task-implementation/run-modes.rst:
```diff
@@ -0,0 +1,211 @@
+Dummy Mode
+----------
+
+**Dummy mode** replaces real jobs with background jobs on the
+scheduler host which use ``sleep`` to simulate the run length according to
+the settings described for simulation mode.
+This avoids :term:`job runner` directives that request compute
+resources for real workflow tasks, and it allows any workflow configuration to
+be run locally in dummy mode.
```
Partially redundant - should be combined with the description higher up?

wxtim replied:
Deliberately somewhat similar - the other text is a top level abstract, this is the introduction to the documentation for dummy mode.

wxtim replied:
Resolved: Fell out of use when I moved the workflow run modes back to the other file

hjoliver commented on src/user-guide/task-implementation/run-modes.rst:
```diff
@@ -0,0 +1,211 @@
+Limitations
+^^^^^^^^^^^
+
+Dummy mode can only be applied to a whole workflow.
```
```suggestion
Simulation mode can only be applied to a whole workflow.
```
(Already said above for Dummy mode).

hjoliver commented on src/user-guide/task-implementation/run-modes.rst:
```diff
@@ -0,0 +1,211 @@
+Skip Mode
+---------
+
+Skip mode is designed as an aid to workflow control:
```
"it allows ..."

```suggestion
Skip mode is designed as an aid to workflow control. It:
```

hjoliver commented on src/user-guide/task-implementation/run-modes.rst:
```diff
@@ -0,0 +1,211 @@
+          house_number = 1..20
+
+We can use skip mode to make sure that a parameter task always
+succeeds without running anything:
```
Nice example :-)

hjoliver commented on src/user-guide/troubleshooting.rst:
```diff
@@ -309,12 +309,12 @@ logged a file in this directory:
 .. code-block:: cylc-graph
 
-   dummy => fetch_files
+   non_functional => fetch_files
```
As noted above, not sure "non-functional" is good terminology.

hjoliver commented on src/user-guide/writing-workflows/scheduling.rst:
```diff
@@ -1301,12 +1301,18 @@ can use a dummy task ``FAM1_done`` like this:
 
+.. note::
+
+   By setting ``run mode = skip`` the task will be follow the
+   :ref:`task-run-modes.skip` submission pathway, which
+   is significantly more efficient than a conventional task.
```
"submission pathway" again. Just say something like this:
```suggestion
   By setting ``run mode = skip`` the task's execution will be simulated
   quickly, with no real job submitted.
```

wxtim commented on src/user-guide/running-workflows/simulation-modes.rst:
```diff
@@ -1,101 +1,30 @@
-   in dummy mode. If your ``init-script`` is required to run even local dummy
-   jobs, the relevant environment setup should be done elsewhere.
+This is useful when developing workflows. Tasks submitted
+using :ref:`task-run-modes.dummy` and
+:ref:`task-run-modes.simulation` pathways can be used while
```
Agreed - lets not show Cylc's guts to the user unless absolutely necessary.

wxtim commented:
I've tried to respond to comments from @MetRonnie and @hjoliver (if you're both saying the same thing I'll run with it).

After a certain amount of thought I've restored the references to dummy _tasks_ and added some disambiguation in the glossary.

I've also renamed the two main files to reflect the difference between workflow and task run modes. To this end https://github.com/cylc/cylc-doc/pull/755/commits/4d9f5355089ab28e9ae99615a861b8881438a041 contains actual changes, whilst https://github.com/cylc/cylc-doc/pull/755/commits/74080029a93bc4ddafe7c9d54c37b51b7fe05abd is just the file renaming.

MetRonnie commented on src/user-guide/running-workflows/workflow-run-modes.rst:
```diff
@@ -0,0 +1,116 @@
+Workflow Run Modes
+------------------
+
+Cylc can run a workflow without running the real jobs, which may be
+long-running and resource-hungry.
+
+This is useful when developing workflows. Tasks submitted
+using :ref:`workflow-run-modes.dummy` and
+:ref:`workflow-run-modes.simulation` can be used while
+developing workflows to understand how different outcomes
+and run lengths will affect the workflow.
```
```suggestion
Cylc can run a workflow without running the real jobs, which may be
useful when developing workflows. Tasks submitted
using :ref:`workflow-run-modes.dummy` and
:ref:`workflow-run-modes.simulation` can be used to understand how different outcomes
and run lengths will affect the workflow.
```

MetRonnie commented on src/user-guide/running-workflows/workflow-run-modes.rst:
```diff
@@ -0,0 +1,116 @@
+developing workflows to understand how different outcomes
+and run lengths will affect the workflow.
+
+To apply a run mode to a workflow:
```
```suggestion
To start a workflow in one of these modes:
```

MetRonnie commented on src/user-guide/running-workflows/workflow-run-modes.rst:
```diff
@@ -0,0 +1,116 @@
+Simulated Failure
+^^^^^^^^^^^^^^^^^
+
+Tasks always complete all custom outputs, by default they succeed.
+
+   If you want to test individual workflow pathways, use
+   :ref:`skip mode <task-run-modes.skip>`.
```
```suggestion
Tasks always complete all custom outputs, and by default they will succeed.

If you want to test individual workflow pathways, use
:ref:`skip mode <task-run-modes.skip>`.
```

MetRonnie commented on src/user-guide/task-implementation/task-run-modes.rst:
```diff
@@ -0,0 +1,112 @@
+.. _task-run-modes:
+
+Task Run Modes
+==============
+
+.. versionadded:: 8.4.0
+
+   * Before 8.4.0 run modes could only be set for the workflow, not
+     for individual tasks.
+   * Skip mode.
+
+
+As well as the default task implementation ("live" mode)
+other implementations can be selected:
```
I think it would be more helpful if this whole page was framed as about skip mode. For the user, I don't think it's useful to talk about "multiple" task run modes; what is important to them is they can set tasks to run in skip mode.

wxtim commented on src/user-guide/troubleshooting.rst:
```diff
@@ -309,12 +309,12 @@ logged a file in this directory:
 If you need to access files from a remote platform (e.g. 2-stage ``fcm_make``),
 ensure that a task has submitted to it before you do so. If needed you can use
-a blank "dummy" task to ensure that remote installation is completed *before*
+a blank "non-functional" task to ensure that remote installation is completed *before*
```
```suggestion
a blank :ref:`dummy task` task to ensure that remote installation is completed *before*
```

hjoliver commented on src/glossary.rst:
```diff
@@ -1767,3 +1767,17 @@ Glossary
+
+      .. important::
+
+         All tasks in a workflow run in dummy or simulation mode are dummy
```
```suggestion
         All tasks that run in dummy or simulation mode are dummy
```

(At first glance I saw "All tasks in a workflow run in dummy or simulation mode" 🤣  )

wxtim replied:
No, it's deliberate? Am I missing something?

hjoliver commented on src/user-guide/running-workflows/workflow-run-modes.rst:
```diff
@@ -0,0 +1,113 @@
+------------------
+
+Cylc can run a workflow without running the real jobs, which may be
+useful when developing workflows. Tasks submitted
+using :ref:`workflow-run-modes.dummy` and
+:ref:`workflow-run-modes.simulation` can be used to understand how different outcomes
```
Sorry if I'm being overly picky, but to my mind tasks don't get "submitted" in simulation mode, because submission is the process of submitting a job to run on the system. I suppose we could spin it that way, but IMO better to avoid even raising the question in user's minds.

```suggestion
useful when developing workflows. Tasks 
in :ref:`workflow-run-modes.dummy` and
:ref:`workflow-run-modes.simulation` can help you to understand how different outcomes
```

wxtim replied:
If you want to be picky you could say that tasks submit jobs and are not themselves submitted. 

I have accepted this, though I'm not sure I agree. I'd describe skip and simulation jobs as being submitted to the scheduler.

hjoliver replied:
> If you want to be picky you could say that tasks submit jobs and are not themselves submitted.

Correct, we shouldn't say that tasks get submitted, just being lazy. They have a submitted state, which means they submitted a job.

> I'd describe skip and simulation jobs as being submitted to the scheduler.

Yeah, that's what I meant by "I suppose we could spin it that way".

But I think it's easy for users to grok if we just say jobs get submitted.

hjoliver commented on src/user-guide/running-workflows/workflow-run-modes.rst:
```diff
@@ -0,0 +1,113 @@
+   $ cylc play --mode=dummy <workflow-id>
+   $ cylc play --mode=simulation <workflow-id>
+
+.. admonition:: Limitations
```
Are these really limitations as such, or just points to be aware of. IMO a limitation is something that you might legitimately expect should work but doesn't.

wxtim replied:
I don't really care what we call it - and to be fair, I think that the first point is something which does fit your definition of a limitation - superficially one might expect to be able to re-run in live mode.

```suggestion
.. note::
```

hjoliver replied:
Fair enough, happy to leave it as a "limitation".

hjoliver commented on src/user-guide/running-workflows/workflow-run-modes.rst:
```diff
@@ -0,0 +1,113 @@
+resources for real workflow tasks, and it allows any workflow configuration to
+be run locally in dummy mode.
+
+.. admonition:: Limitations
```
(In this case I agree, these are limitations ... sensible, but limitations nonetheless)

hjoliver commented on src/user-guide/running-workflows/workflow-run-modes.rst:
```diff
@@ -0,0 +1,113 @@
+---------------
+
+**Simulation mode** does not run real jobs, and does not generate job
+log files.
```
```suggestion
log files.  Instead, the scheduler internally simulates task completion to evolve
the workflow.
```

hjoliver commented on src/user-guide/task-implementation/task-run-modes.rst:
```diff
@@ -0,0 +1,127 @@
+
+Skip mode is designed as an aid to workflow control:
+
+* Allows creation of :term:`dummy tasks<dummy task>` as part of workflow
+  design.
+* Allows skipping of tasks in a running workflow using
+  ``cylc broadcast -s 'run mode = skip'``.
```
```suggestion
* It allows creation of :term:`dummy tasks<dummy task>` as part of workflow
  design.
* It allows skipping of tasks in a running workflow using
  ``cylc broadcast -s 'run mode = skip'``.
```

hjoliver commented on src/user-guide/task-implementation/task-run-modes.rst:
```diff
@@ -0,0 +1,127 @@
+        * example_task
+
+   This is designed to prevent users running a task in skip mode by mistake.
+   If you are using skip mode as part of a workflow design then this can be
+   ignored.
```
```suggestion
   If you are using skip mode deliberately then this can be
   ignored.
```
(minor, but more concise)

hjoliver commented on src/user-guide/task-implementation/task-run-modes.rst:
```diff
@@ -0,0 +1,127 @@
+* All required outputs will be generated.
+* Succeeded will be generated if success is optional.
+* If outputs is specified and does not include either succeeded
+  or failed then succeeded will be produced.
```
```suggestion
  or failed then succeeded will be generated.
```
(consistency)

hjoliver commented on src/user-guide/task-implementation/task-run-modes.rst:
```diff
@@ -0,0 +1,127 @@
+* If outputs is specified and does not include either succeeded
+  or failed then succeeded will be produced.
+
+The outputs submitted and started are always produced and do not
```
```suggestion
The outputs submitted and started are always generated and do not
```

hjoliver commented on src/user-guide/task-implementation/task-run-modes.rst:
```diff
@@ -0,0 +1,127 @@
+
+.. note::
+
+   Skip mode only works on "live workflows". Workflows run in
```
```suggestion
   Skip mode only works on live workflows. Workflows run in
```
(quotes not needed there?)

hjoliver commented on src/user-guide/task-implementation/task-run-modes.rst:
```diff
@@ -0,0 +1,127 @@
+
+.. admonition:: Scenario
+
+   We don't need a task to run, but want to set it to succeed.
```
```suggestion
   We don't need a task to run, but want to set it to succeeded.
```
(I suppose "set it to succeed" could be OK, but to me that sounds kinda like pre-setting it somehow to succeed when it does run - which it won't do).

hjoliver reviewed (commented):
Looking good now, generally, just a bunch of minor suggestions.

hjoliver reviewed (approved):
Nice. I like your glossary entry on "dummy task".

MetRonnie commented on src/user-guide/running-workflows/workflow-run-modes.rst:
```diff
@@ -0,0 +1,114 @@
+
+Tasks always complete all custom outputs, and by default they will succeed.
+
+If you want to test individual workflow pathways, use
+:ref:`skip mode <task-run-modes.skip>`.
```
Sorry, what does this mean?

wxtim replied:
If you have a case:

```
a:foo? => f
b:bar? => b
```

Simulation mode will always run f & b, but if you want to simulate only emitting one of them then your only option at present is skip mode.

How about

```suggestion
If you want to test emitting only some custom outputs
you can run the workflow in live mode with task run modes
set to :ref:`skip mode <task-run-modes.skip>`.
```

MetRonnie commented on src/user-guide/task-implementation/task-run-modes.rst:
```diff
@@ -0,0 +1,127 @@
+can be enabled using
+:cylc:conf:`[runtime][<namespace>][skip]disable task event handlers`.
+
+.. note::
+
+   Skip mode only works on live workflows. Workflows run in
+   :ref:`workflow-run-modes.simulation`
+   and :ref:`workflow-run-modes.dummy` ignore the
+   :cylc:conf:`[runtime][<namespace>]run mode` setting.
```
This note feels out of place under this "task event handlers" subsection?

wxtim replied:
```suggestion
```

MetRonnie commented on src/user-guide/task-implementation/task-run-modes.rst:
```diff
@@ -0,0 +1,127 @@
+   :cylc:conf:`[runtime][<namespace>]run mode` setting.
+
+
+Set task to succeeded
+^^^^^^^^^^^^^^^^^^^^^
```
Feel like this should come under an "examples" section

MetRonnie commented on src/user-guide/task-implementation/task-run-modes.rst:
```diff
@@ -0,0 +1,127 @@
+
+   We don't need a task to run, but want to set it to succeeded.
+
+Broadcast :cylc:conf:`[runtime][<namespace>]run mode` setting the
+value to ``skip``.
```
Or can use `cylc set --out skip`?

oliver-sanders replied:
This use case needs fleshing out, we're not actually interested in how to set a task to succeeded here.

The use case is along the lines of, I want to turn a task off in my workflow, or in a particular cycle of my workflow. So I don't want to set the output ahead of time (`cylc set --out=succeeded` or skip or whatever) because that would cause downstream triggering (although I could use the `--wait` option...), so I configure the task not to run instead.

MetRonnie commented on src/user-guide/troubleshooting.rst:
```diff
@@ -309,7 +309,7 @@ logged a file in this directory:
 If you need to access files from a remote platform (e.g. 2-stage ``fcm_make``),
 ensure that a task has submitted to it before you do so. If needed you can use
-a blank "dummy" task to ensure that remote installation is completed *before*
+a blank :term:`dummy task` task to ensure that remote installation is completed *before*
```
```suggestion
a blank :term:`dummy task` to ensure that remote installation is completed *before*
```

MetRonnie commented on src/user-guide/running-workflows/workflow-run-modes.rst:
```diff
@@ -0,0 +1,114 @@
+
+.. admonition:: Limitations
+
+   * Workflow run mode is recorded in the workflow run database. Cylc will not let you
+     *restart* a dummy mode workflow in live mode, or vice versa. Instead,
```
```suggestion
   * A workflow cannot be :term:`restarted <restart>` in a different run mode. Instead,
```

MetRonnie commented on src/user-guide/running-workflows/workflow-run-modes.rst:
```diff
@@ -0,0 +1,114 @@
+If :cylc:conf:`[runtime][<namespace>]execution time limit` and
+:cylc:conf:`[runtime][<namespace>][simulation]speedup factor` are both set,
+run length is computed by dividing the
+:cylc:conf:`[runtime][<namespace>]execution time limit` by
+:cylc:conf:`[runtime][<namespace>][simulation]` speedup factor.
```
```suggestion
execution time limit by the speedup factor.
```

MetRonnie commented on src/user-guide/running-workflows/workflow-run-modes.rst:
```diff
@@ -0,0 +1,114 @@
+:ref:`skip mode <task-run-modes.skip>`.
+
+You can set some or all instances of a task to fail using
+:cylc:conf:`[runtime][<namespace>][simulation]fail cycle points`.
+``fail cycle points`` takes either a list of cycle point strings or "all".
```
```suggestion
:cylc:conf:`[runtime][<namespace>][simulation]fail cycle points`,
which takes either a list of cycle point strings or "all".
```

MetRonnie commented on src/user-guide/task-implementation/task-run-modes.rst:
```diff
@@ -0,0 +1,127 @@
+
+.. versionadded:: 8.4.0
+
+   * Before 8.4.0 run modes could only be set for the workflow, not
+     for individual tasks.
+   * Skip mode.
+
+Skip mode is designed as an aid to workflow control:
+
+* It allows creation of :term:`dummy tasks<dummy task>` as part of workflow
+  design.
+* It allows skipping of tasks in a running workflow using
+  ``cylc broadcast -s 'run mode = skip'``.
```
```suggestion
Skip mode is designed as an aid to workflow control:

* It allows creation of :term:`dummy tasks<dummy task>` as part of workflow
  design.
* It allows skipping of tasks in a running workflow using either:
  * ``cylc broadcast -s 'run mode = skip'``
  * ``cylc set --out skip``
```

oliver-sanders commented on src/user-guide/task-implementation/task-run-modes.rst:
```diff
@@ -0,0 +1,127 @@
+Broadcast :cylc:conf:`[runtime][<namespace>]run mode` setting the
+value to ``skip``.
+
+Create a Graph Control Task
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. admonition:: Scenario
+
+   We have a large family to large family trigger.
+
+   .. code-block:: cylc
+
+      [task parameters]
+          x = 1..N   # Where N is big.
+      [scheduling]
+          [[graph]]
+              R1 = FAMILY:succeed-all => FAMILY2
+      [runtime]
+         [[FAMILY, FAMILY2]]
+         [[task1<x>]]
+            inherit = FAMILY
+         [[task2<x>]]
+            inherit = FAMILY2
+
+   If we increase N, the number of dependencies being tracked
+   by the scheduler is N^2.
+
+Adding a skip mode task between the families:
```
This is duplicating the workflow design guide (which hasn't been updated for skip mode).

Please cover this topic here: https://cylc.github.io/cylc-doc/stable/html/workflow-design-guide/efficiency.html#family-to-family-triggering

wxtim replied:
Which itself is something of a duplication of https://cylc.github.io/cylc-doc/stable/html/user-guide/writing-workflows/scheduling.html#efficient-inter-family-triggering

How about I put up an issue to merge the two and just put a reference in one place?

wxtim commented on src/glossary.rst:
```diff
@@ -1767,3 +1767,17 @@ Glossary
 
+   dummy task
+      A task which runs a trivially simple script such as ``sleep 1``,
+      ``exit 0`` or ``true``, or which uses alternative :ref:`task-run-modes.skip`
+      or :ref:`workflow-run-modes` to run such a script or avoid running a
+      script at all.
+
+      .. important::
+
+         All tasks that run in dummy or simulation mode are dummy
+         tasks, but the methods outlined above will create dummy tasks in a
+         live workflow.
```
I'm on the fence about this. 

It makes graph focused dummy task totally separate from dummy _mode_. However the post change doesn't rule out dummy mode tasks being dummy tasks (since dummy mode essentially sets `script = true` for all tasks). So I think it's better to keep describing tasks in dummy and simulation mode as a subset.

MetRonnie replied:
I'm committing this suggestion as I think this is the one you mentioned in person? (IMO the concept of dummy tasks only makes sense in live mode; there is no need to make a distinction about if tasks are dummy tasks or not in dummy/simulation mode - leave that to the philosophers)

wxtim commented on src/user-guide/task-implementation/task-run-modes.rst:
```diff
@@ -0,0 +1,127 @@
+   :ref:`workflow-run-modes.simulation`
+   and :ref:`workflow-run-modes.dummy` ignore the
+   :cylc:conf:`[runtime][<namespace>]run mode` setting.
+
```
```suggestion
```

MetRonnie commented on src/user-guide/task-implementation/task-run-modes.rst:
Might be better to rename this to `skip-mode.rst`?

MetRonnie commented on src/user-guide/task-implementation/task-run-modes.rst:
```diff
@@ -0,0 +1,119 @@
+
+.. note::
+
+   Setting :cylc:conf:`[runtime][<namespace>]run mode` in your ``flow.cylc``
```
```suggestion
   Setting :cylc:conf:`[runtime][<namespace>]run mode=skip` in your ``flow.cylc``
```

MetRonnie commented on src/user-guide/task-implementation/task-run-modes.rst:
```diff
@@ -0,0 +1,119 @@
+can be enabled using
+:cylc:conf:`[runtime][<namespace>][skip]disable task event handlers`.
+
+Skip Mode Examples
+------------------
```
This isn't the right heading level

wxtim replied:
Is it not? Or is our theme not displaying headings right?

https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#sections

Happy to bump it up a level.

wxtim replied:
```suggestion
Skip Mode Examples
==================
```

MetRonnie commented on src/user-guide/interventions/skip-cycle.gui.gif:
Can't help but think Edit Runtime would be simpler ?

wxtim replied:
Much easier, but I'd missed that you could do it to the root task. Nice.

oliver-sanders replied:
Suggest picking a cycle point in the 2000's to make it clearer (users are more used to datetime cycle points).

I'm not sure what the `set` at the end did?

wxtim replied:
> Suggest picking a cycle point in the 2000's to make it clearer (users are more used to datetime cycle points).

Went for 1900 before this comment.

> I'm not sure what the `set` at the end did?

Triggered the cycle I'd set to skip mode.

oliver-sanders replied:
> Went for 1900 before this comment.

:+1:

>> I'm not sure what the set at the end did?
>
> Triggered the cycle I'd set to skip mode.

It's not so obvious that setting task `z` in one cycle causes task `x` to run in the next.

Suggest unpausing the workflow instead to make it clear that no further intervention is required after the broadcast?

oliver-sanders commented on src/user-guide/interventions/index.rst:
```diff
@@ -488,3 +488,37 @@ Hold The Workflow And Trigger Tasks One By One
+
+      .. code-block:: console
+
+         cylc broadcast -p '<cycle-glob>' -n root -s 'run mode = skip'
```
Cylc broadcast doesn't actually accept globs. It does except the `*` wildcard, but you can't actually glob with it (e.g. `2000*`).

Maybe add to the note explaining that `root` matches all tasks, but that this could be swapped out for more targeted interventions.

oliver-sanders commented on src/user-guide/interventions/index.rst:
```diff
@@ -488,3 +488,38 @@ Hold The Workflow And Trigger Tasks One By One
+
+.. note::
+
+   `-n root` matches all tasks in a cycle. Similarly, it is possible to
```
@wxtim, will fix the test failure

```suggestion
   ``-n root`` matches all tasks in a cycle. Similarly, it is possible to
```

oliver-sanders commented on src/user-guide/running-workflows/workflow-run-modes.rst:
```diff
@@ -0,0 +1,113 @@
+.. _workflow-run-modes:
+
+Workflow Run Modes
+------------------
```
```suggestion
==================
```

oliver-sanders commented on src/user-guide/task-implementation/skip-mode.rst:
```diff
@@ -0,0 +1,119 @@
+.. _task-run-modes.skip:
+
+Skip Mode
+=========
```
Heading levels are a bit squiffy in this file.

oliver-sanders commented on src/user-guide/task-implementation/skip-mode.rst:
```diff
@@ -0,0 +1,119 @@
+* It allows creation of :term:`dummy tasks<dummy task>` as part of workflow
+  design.
+* It allows skipping of tasks in a running workflow using either:
+  * ``cylc broadcast -s 'run mode = skip'`` (for when it is ready to run)
+  * ``cylc set --out skip`` (to immediately skip)
```
Note, `broadcast` works with any task or family, whereas `cylc set` only works with `n=0` tasks or tasks with an explicit ID (excluding families or globs).

oliver-sanders commented on src/user-guide/task-implementation/skip-mode.rst:
```diff
@@ -0,0 +1,119 @@
+
+* All required outputs will be generated.
+* ``succeeded`` will be generated even if success is :term:`optional <optional output>`.
+* If ``[skip]outputs`` is specified and does not include either ``succeeded``
```
This will insert a link to the config, please test.

```suggestion
* If `[skip]outputs <flow.cylc[runtime][<namespace>][skip]outputs>` is specified and does not
   include either ``succeeded``
```

oliver-sanders commented on src/user-guide/task-implementation/skip-mode.rst:
```diff
@@ -0,0 +1,119 @@
+
+By default task event handlers are disabled by skip mode, but they
+can be enabled using
+:cylc:conf:`[runtime][<namespace>][skip]disable task event handlers`.
```
Referencing long configurations can be a pain when writing the docs and full configuration paths might be overkill if the context of the writing is obvious.

FYI, you can shorten these absolute configs to relative ones in one of two ways...

1. We can use the `.. cylc:scope::` directive, all future `:cylc:conf:` references will be relative to the specified section. Remember to reset this later on in the file:

```rest
.. cylc:scope:: flow.cylc[runtime][<namespace>]

can be enabled by using :cylc:conf:`[skip]disable task event handlers`.

.. cylc:scope::
```

2. Use a `<target>`

```rest
:cylc:conf:`[skip]disable task event handlers <flow.cylc[runtime][<namespace>][skip]disable task event handlers>`

wxtim replied:
Thanks. TIL - the former seems handy in this context.

wxtim replied:
I couldn't get the `.. cylc:scope::` directive working as shown

wxtim replied:
NVM

```diff
- .. cylc:scope::
+ .. cylc-scope::
```

oliver-sanders commented on src/user-guide/task-implementation/skip-mode.rst:
```diff
@@ -0,0 +1,119 @@
+Broadcast :cylc:conf:`[runtime][<namespace>]run mode` setting the
+value to ``skip``.
+
+.. code-block::console
```
I think this code block will be treated as a comment due to this missing space:

```suggestion
.. code-block:: console
```

oliver-sanders reviewed (approved):
Approved with fixes.
